File: leetcode_javascript/560.SubarraySumEqualsK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    const n = nums.length;
    let sum = 0, count = 0;
    const map = new Map(); // num, count

    map.set(0, 1);  // don't forget.
    nums.forEach(num => {
        sum += num;
        const ct = map.get(sum - k) || 0;
        if (ct > 0) {
            count += ct;
        }
        map.set(sum, (map.get(sum) || 0) + 1);
    });
    return count;
};

File: leetcode/560.SubarraySumEqualsK.cpp
// Space O(n), Time O(n)
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int count = 0, sum = 0;
        map<int, int> mp;
        mp[0] = 1; // 1 subarray that sums to 0. 1, NOT -1 !!!
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            if (mp.contains(sum - k)) {
                count += mp[sum - k];
            }
            ++ mp[sum];
        }
        return count;
    }
};

/**
560. Subarray Sum Equals K

Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.

A subarray is a contiguous non-empty sequence of elements within an array.
 */
